crates/telio-starcast/src/multicaster.rs

```diff
@@ -0,0 +1,295 @@
+const MULTICAST_TRANSPORT_PORT: u16 = 0;
+#[cfg(not(feature = "test-util"))]
+const MULTICAST_TRANSPORT_PORT: u16 = 13569;
+const MAX_PACKET: usize = 2048;
+const IPV4_HEADER: usize = 20; // bytes
+const UDP_DEST_OFFSET: usize = 2;
+
+/// Multicaster task that resends locally generated
+/// multicast traffic to other peers
+pub struct Multicaster {
+    wg_if: Arc<DynamicWg>,
+    transport_socket: Arc<UdpSocket>,
+    local_socket: Arc<UdpSocket>,
+    wg_sock: Arc<SocketAddr>,
+    chan: Arc<RwLock<Chan<Vec<u8>>>>,
+    tun_to_transport_task: Option<JoinHandle<()>>,
+    transport_to_local_task: Option<JoinHandle<()>>,
+    peer: Arc<Tunn>,
+    nat_map: Arc<RwLock<HashMap<u16, SocketAddr>>>,
+    multicaster_ip: Arc<MulticasterIp>,
+}
+
+impl Multicaster {
+    pub async fn new(
+        local_socket: Arc<UdpSocket>,
+        wg: Arc<DynamicWg>,
+        wg_sock: Arc<SocketAddr>,
+        socket_pool: Arc<SocketPool>,
+        chan: Chan<Vec<u8>>,
+        peer: Arc<Tunn>,
+        multicaster_ip: Arc<MulticasterIp>,
+    ) -> Result<Self, String> {
+        let transport_socket = socket_pool
+            .new_internal_udp(
+                SocketAddr::from(([0, 0, 0, 0], MULTICAST_TRANSPORT_PORT)),
+                None,
+            )
+            .await
+            .map_err(|e| {
+                telio_log_error!("Failed to bind multicast socket: {:?}", e);
+                format!("Failed to bind multicast socket: {:?}", e)
+            })?;
+
+        let nat_map: Arc<RwLock<HashMap<u16, SocketAddr>>> = Arc::new(RwLock::new(HashMap::new()));
+
+        Ok(Self {
+            wg_if: wg,
+            transport_socket: Arc::new(transport_socket),
+            local_socket,
+            wg_sock,
+            chan: Arc::new(RwLock::new(chan)),
+            tun_to_transport_task: None,
+            transport_to_local_task: None,
+            peer: peer.clone(),
+            nat_map: nat_map.clone(),
+            multicaster_ip,
+        })
+    }
+
+    pub async fn start_multicaster(&mut self) {
+        self.tun_to_transport_task = Some(tokio::spawn(Multicaster::tun_to_transport(
+            self.chan.clone(),
+            self.wg_if.clone(),
+            self.transport_socket.clone(),
+            self.nat_map.clone(),
+            self.multicaster_ip.clone(),
+        )));
+        self.transport_to_local_task = Some(tokio::spawn(Self::transport_to_local(
+            self.transport_socket.clone(),
+            self.local_socket.clone(),
+            self.wg_sock.clone(),
+            self.peer.clone(),
+            self.nat_map.clone(),
+            self.multicaster_ip.clone(),
+        )));
+    }
+
+    pub async fn stop(&mut self) {
+        if let Some(handle) = &self.transport_to_local_task {
+            handle.abort();
+        }
+        if let Some(handle) = &self.tun_to_transport_task {
+            handle.abort();
+        }
+    }
+
+    pub async fn tun_to_transport(
+        read_chan: Arc<RwLock<Chan<Vec<u8>>>>,
```

Comment:
This is a heavy anti patten, you should never have channel in RwLock, this causes often leads to deadlocks.